src/interfaces/records/handlers/records-upload.ts

```diff
@@ -0,0 +1,196 @@
+import type { MethodHandler } from '../../types.js';
+import type { DidResolver, MessageStore, UploadStore } from '../../../index.js';
+import type { RecordsUploadCompleteMessage, RecordsUploadPartMessage, RecordsUploadStartMessage } from '../types.js';
+
+import { authenticate } from '../../../core/auth.js';
+import { MessageReply } from '../../../core/message-reply.js';
+import { RecordsUpload } from '../messages/records-upload.js';
+import { DwnInterfaceName, DwnMethodName, DwnStateName } from '../../../core/message.js';
+
+type RecordsUploadMessageVariant = RecordsUploadCompleteMessage | RecordsUploadPartMessage | RecordsUploadStartMessage;
+
+export class RecordsUploadHandler implements MethodHandler {
+
+  constructor(private didResolver: DidResolver, private messageStore: MessageStore, private uploadStore: UploadStore) { }
+
+  public async handle({
+    tenant,
+    message,
+    dataStream
+  }): Promise<MessageReply> {
+    const incomingMessage = message as RecordsUploadMessageVariant;
+
+    let recordsUpload: RecordsUpload;
+    try {
+      recordsUpload = await RecordsUpload.parse(incomingMessage);
+    } catch (e) {
+      return new MessageReply({
+        status: { code: 400, detail: e.message }
+      });
+    }
+
+    // authentication & authorization
+    try {
+      await authenticate(incomingMessage.authorization, this.didResolver);
+      await recordsUpload.authorize(tenant, this.messageStore);
+    } catch (e) {
+      return new MessageReply({
+        status: { code: 401, detail: e.message }
+      });
+    }
+
+    // get existing upload start messages matching the `recordId`
+    const query = {
+      interface : DwnInterfaceName.Records,
+      method    : DwnMethodName.Upload,
+      recordId  : incomingMessage.recordId
+    };
+    const existingMessages = await this.messageStore.query(tenant, query) as RecordsUploadMessageVariant[];
+
+    try {
+      const existingUploadComplete = await RecordsUpload.getUploadComplete(existingMessages);
+      if (existingUploadComplete) {
+        return new MessageReply({
+          status: { code: 400, detail: 'upload already complete' }
+        });
+      }
+    } catch (e) {
+      // ignore error if upload complete cannot be found
+    }
+
+    const newMessageIsUploadStart = await recordsUpload.isUploadStart();
+    const newMessageIsUploadComplete = await recordsUpload.isUploadComplete();
+
+    const parts = new Set<number>();
+
+    try {
+      const existingUploadStart = await RecordsUpload.getUploadStart(existingMessages);
+      if (existingUploadStart) {
```

Comment:
`getUploadStart` throws when there is no `existingUploadStart`, so this won't address the valid case where `newMessageIsUploadStart` and there is no `existingUploadStart`.

I mentioned in another comment above that it's a bit odd that the fact that `getUploadComplete` throws is odd to me. Same goes for `getUploadStart`. Personal preference for returning `undefined`.